### PR TITLE
CORE-13560 - initial crypto metrics per operation type

### PR DESCRIPTION
**Pull request by conal-smith**

### CODE REVIEW

**Review by dickon — DISMISSED** [review]
LGTM

**Review by rick-r3 — COMMENTED** [review]
Makes sense.

**Comment by kyriathar** [comment]
Is this conflicting with [#3874](https://github.com/corda/corda-runtime-os/pull/3874)? Should that be merged first?

**Review by ronanbrowne — APPROVED** [review]
LGTM - build

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt [hunk 1]
```diff
@@ -15,32 +16,37 @@ fun getEntityManagerFactory(
...
+                        ?: throw IllegalStateException(
+                            "persistenceUnitName ${CordaDb.Crypto.persistenceUnitName} is not registered."
+                        )
                 )
-        )
-    }
-    return entityManagerFactory
+            }
+            entityManagerFactory
+        }!!
```
Comment (dickon) [inline components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:51]: We should not use !! in production code

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt [hunk 2]
```diff
@@ -15,32 +16,37 @@ fun getEntityManagerFactory(
...
-                    tenantId
-                )
-            )?.cryptoDmlConnectionId
-                ?: throw IllegalStateException(
-                    "virtual node for $tenantId is not registered."
-                ),
-            entitiesSet = jpaEntitiesRegistry.get(CordaDb.Crypto.persistenceUnitName)
-                ?: throw IllegalStateException(
-                    "persistenceUnitName ${CordaDb.Crypto.persistenceUnitName} is not registered."
+    return CordaMetrics.Metric.EntityManagerFactoryCreationTimer.builder()
```
Comment (dickon) [inline components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:19]: better to use an expression body here

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt (hunk 1 above)
Comment (dickon) [reply components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:51]: Also, there's no test coverage on this metric, or the others.

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt (hunk 2 above)
Comment (dickon) [reply components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:19]: Please instrument both sides of the if(onCluster) separately so we can see what's going on. In particular, the true branch shouldn't be taking much time more than once.

Comment (dickon) [reply components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:19]: Minor point, and probably a bit late now, but I'd rather see metrics like this imlemented as function annotations, so we aren't adding unnecessarily deep block nesting all over the place. For one, the change of block structure makes patches like this harder to review than it needs to be.

**Review by dickon — APPROVED** [review]
Okay as far as it goes; there's certainly more we can do, so approved for now, and these are ways to take this further:

We are missing a metric on `SoftCryptoServiceProviderImpl.getInstance`. It's probably nothing but I want to be sure the caching with the `lazy` works.

We should measure the SigningSerivceImpl.getOwnedKeyRecord, separately for each branch fo the `if`.

It's probably trivial, but worth measuring `SoftCryptoServicePRoviderImpl.decodePublicKey` and `CryptoSignatureSpec.toSignatureSpec` 

`SigningRepositoryFacotryImpl.getInstance`  and `CryptoServiceFacotryImpl.getInstanc` makes an EMF so is worth measuring.

`SoftCryptoServiceProviderImpl.getInstance` is expensive so let's measure that.


A few general comments:

- We generally avoid !! in production code
- There's no tests, i.e. nothing in the CI that proves these metrics do what we think they do.

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt (hunk 2 above)
Comment (sprintengine) [reply components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:19]: Thanks, I'll make some further updates in a later PR to address some of these comments and add more metrics. Just this one is next in line to get merged asap

**Pull request merged**